File: apple/Nianxiang/Particle/ParticleView.swift
/// PUBLIC STUB — the real ParticleEngine is a Metal compute-shader particle system
/// (spring/curl physics, bokeh, sand-burst, text dissolve) and lives in the private
/// core module. This stub keeps the exact public API and shows the submitted photo
/// in a plain platform image view, so the app remains fully usable without the
/// particle effects. Gesture/camera methods are no-ops (there is no 3D camera).
import Foundation
import SwiftUI
#if os(macOS)
import AppKit
#else
import UIKit
#endif

@MainActor
protocol ParticleEngineListener: AnyObject {
    func onPhotoPrepared(photoId: String)
    func onEntranceStarted(photoId: String)
    func onRenderError(message: String)
    func onPerformanceSample(_ sample: ParticlePerformanceSample)
}

extension ParticleEngineListener {
    func onPhotoPrepared(photoId: String) {}
    func onEntranceStarted(photoId: String) {}
    func onRenderError(message: String) {}
    func onPerformanceSample(_ sample: ParticlePerformanceSample) {}
}

@MainActor
final class ParticleEngine {
    weak var listener: (any ParticleEngineListener)?

    #if os(macOS)
    let view = NSImageView()
    #else
    let view = UIImageView()
    #endif

    private let worker = DispatchQueue(label: "particle-stub-decode", qos: .utility)
    private var shownPhotoId: String?
    private var decodingPhotoId: String?

    init() {
        #if os(macOS)
        view.imageScaling = .scaleProportionallyUpOrDown
        view.wantsLayer = true
        view.layer?.backgroundColor = NSColor.black.cgColor
        #else
        view.contentMode = .scaleAspectFill
        view.clipsToBounds = true
        view.backgroundColor = .black
        view.isUserInteractionEnabled = false
        #endif
    }

    func submit(_ scene: ParticleSceneState) {
        guard let photoId = scene.photoId else {
            shownPhotoId = nil
            decodingPhotoId = nil
            view.image = nil
            return
        }
        guard photoId != shownPhotoId, photoId != decodingPhotoId, let jpeg = scene.jpeg else { return }
        decodingPhotoId = photoId
        worker.async { [weak self] in
            #if os(macOS)
            let image = NSImage(data: jpeg)
            #else
            let image = UIImage(data: jpeg)
            #endif
            Task { @MainActor [weak self] in
                guard let self, self.decodingPhotoId == photoId else { return }
                self.decodingPhotoId = nil
                guard let image else {
                    self.listener?.onRenderError(message: "photo decode failed")
                    return
                }
                self.view.image = image
                self.shownPhotoId = photoId
                self.listener?.onPhotoPrepared(photoId: photoId)
                self.listener?.onEntranceStarted(photoId: photoId)
            }
        }
    }

    func setTestConfiguration(particleBudget: Int, seed: Int?, animationTimeSeconds: Float? = nil) {}

    func orbitBy(deltaX: Float, deltaY: Float, screenX: Float, screenY: Float) {}
    func zoomBy(scaleFactor: Float) {}
    func pulseAt(screenX: Float, screenY: Float) {}
    func resetCamera() {}
    func clearParticleFocus() {}

    func release() {
        shownPhotoId = nil
        decodingPhotoId = nil
        view.image = nil
    }
}

// MARK: - SwiftUI wrapper

#if os(macOS)
struct ParticleCanvas: NSViewRepresentable {
    let engine: ParticleEngine

    func makeNSView(context: Context) -> NSImageView { engine.view }
    func updateNSView(_ nsView: NSImageView, context: Context) {}
}
#else
struct ParticleCanvas: UIViewRepresentable {
    let engine: ParticleEngine

    func makeUIView(context: Context) -> UIImageView { engine.view }
    func updateUIView(_ uiView: UIImageView, context: Context) {}
}
#endif

File: android/app/src/main/java/com/nianxiang/app/particle/ParticleView.kt
/** PUBLIC STUB — the real ParticleView is an OpenGL ES 3.1 compute-shader particle
 *  system (spring/curl physics, bokeh, sand-burst, text dissolve) and lives in the
 *  private core module. This stub keeps the exact public API and draws the submitted
 *  photo as a static cover-fit bitmap, so the app remains fully usable without the
 *  particle effects. Gesture/camera methods are no-ops (there is no 3D camera). */
package com.nianxiang.app.particle

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import java.util.concurrent.Executors

class ParticleView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
) : View(context, attrs) {
    interface Listener {
        fun onPhotoPrepared(photoId: String) = Unit
        fun onEntranceStarted(photoId: String) = Unit
        fun onRenderError(message: String) = Unit
        fun onPerformanceSample(sample: ParticlePerformanceSample) = Unit
    }

    var listener: Listener? = null

    private val worker = Executors.newSingleThreadExecutor { task ->
        Thread(task, "particle-stub-decode").apply { priority = Thread.NORM_PRIORITY - 1 }
    }
    private val paint = Paint(Paint.FILTER_BITMAP_FLAG)
    private var bitmap: Bitmap? = null
    private var shownPhotoId: String? = null
    private var decodingPhotoId: String? = null

    fun submit(scene: ParticleSceneState) {
        val photoId = scene.photoId
        if (photoId == null) {
            shownPhotoId = null
            decodingPhotoId = null
            bitmap = null
            invalidate()
            return
        }
        if (photoId == shownPhotoId || photoId == decodingPhotoId || scene.jpeg == null) return
        decodingPhotoId = photoId
        val jpeg = scene.jpeg
        worker.execute {
            val decoded = runCatching { decodeCapped(jpeg) }.getOrNull()
            post {
                if (decodingPhotoId != photoId) return@post
                decodingPhotoId = null
                if (decoded == null) {
                    listener?.onRenderError("photo decode failed")
                    return@post
                }
                bitmap = decoded
                shownPhotoId = photoId
                invalidate()
                listener?.onPhotoPrepared(photoId)
                listener?.onEntranceStarted(photoId)
            }
        }
    }

    /** cap the decode around 2048px on the long side — it is only ever drawn once */
    private fun decodeCapped(jpeg: ByteArray): Bitmap? {
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeByteArray(jpeg, 0, jpeg.size, bounds)
        var sample = 1
        while (maxOf(bounds.outWidth, bounds.outHeight) / (sample * 2) >= 2048) sample *= 2
        val opts = BitmapFactory.Options().apply { inSampleSize = sample }
        return BitmapFactory.decodeByteArray(jpeg, 0, jpeg.size, opts)
    }

    fun setTestConfiguration(
        particleBudget: Int,
        seed: Int?,
        animationTimeSeconds: Float? = null,
    ) {
        // no renderer in the public build
    }

    fun orbitBy(deltaX: Float, deltaY: Float, screenX: Float, screenY: Float) = Unit
    fun zoomBy(scaleFactor: Float) = Unit
    fun pulseAt(screenX: Float, screenY: Float) = Unit
    fun resetCamera() = Unit
    fun clearParticleFocus() = Unit

    fun release() {
        worker.shutdownNow()
        bitmap = null
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawColor(Color.BLACK)
        val image = bitmap ?: return
        val scale = maxOf(width.toFloat() / image.width, height.toFloat() / image.height)
        val matrix = Matrix().apply {
            setScale(scale, scale)
            postTranslate((width - image.width * scale) / 2f, (height - image.height * scale) / 2f)
        }
        canvas.drawBitmap(image, matrix, paint)
    }
}
